README.md
# relative-paths [![NPM version](https://img.shields.io/npm/v/clean-stacktrace-relative-paths.svg?style=flat)](https://www.npmjs.com/package/clean-stacktrace-relative-paths) [![mit license][license-img]][license-url] [![NPM monthly downloads](https://img.shields.io/npm/dm/clean-stacktrace-relative-paths.svg?style=flat)](https://npmjs.org/package/clean-stacktrace-relative-paths) [![npm total downloads][downloads-img]][downloads-url]

> Meant to be used with [clean-stacktrace][] as mapper function. Makes absolute paths inside stack traces to relative paths.

[![code climate][codeclimate-img]][codeclimate-url] 
[![code style][standard-img]][standard-url] 
[![linux build][travis-img]][travis-url] 
[![windows build][appveyor-img]][appveyor-url] 
[![code coverage][coverage-img]][coverage-url] 
[![dependency status][david-img]][david-url]
[![paypal donate][paypalme-img]][paypalme-url] 

You might also be interested in [always-done](https://github.com/hybridables/always-done#readme).

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [API](#api)
  * [relativePaths](#relativepaths)
- [Related](#related)
- [Contributing](#contributing)
- [Building docs](#building-docs)
- [Running tests](#running-tests)
- [Author](#author)
- [License](#license)

_(TOC generated by [verb](https://github.com/verbose/verb) using [markdown-toc](https://github.com/jonschlinkert/markdown-toc))_

## Install
Install with [npm](https://www.npmjs.com/)

```
$ npm install clean-stacktrace-relative-paths --save
```

or install using [yarn](https://yarnpkg.com)

```
$ yarn add clean-stacktrace-relative-paths
```

## Usage
> For more use-cases see the [tests](test.js)

```js
const cleanStacktraceRelativePaths = require('clean-stacktrace-relative-paths')
```

## API

### [relativePaths](index.js#L51)
> Cleans stack traces to use relative paths, instead of absolute paths. Pass a `cwd` if you want to customize, otherwise `process.cwd()` is used by default.

Meant to be used as `mapper` argument in [clean-stacktrace][]

**Params**

* `[cwd]` **{String}**: custom current working directory    
* `returns` **{Function}**: a function that can be passed as `mapper` to [clean-stacktrace][]  

**Example**

```js
const cleanStack = require('clean-stacktrace')
const relativePaths = require('clean-stacktrace-relative-paths')

const error = new Error('Missing unicorn')
console.log(error.stack)
// =>
// Error: Missing unicorn
//     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)
//     at Module._compile (module.js:409:26)
//     at Object.Module._extensions..js (module.js:416:10)
//     at Module.load (module.js:343:32)
//     at Function.Module._load (module.js:300:12)
//     at Function.Module.runMain (module.js:441:10)
//     at startup (node.js:139:18)

const mapper = relativePaths()
const stack = cleanStack(error.stack, mapper)
console.log(stack)
// =>
// Error: Missing unicorn
//     at Object.<anonymous> (unicorn.js:2:15)
```

## Related
- [always-done](https://www.npmjs.com/package/always-done): Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A drop-in replacement… [more](https://github.com/hybridables/always-done#readme) | [homepage](https://github.com/hybridables/always-done#readme "Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A drop-in replacement for [async-done][] - pass 100% of its tests plus more")
- [assert-kindof](https://www.npmjs.com/package/assert-kindof): Check native type of value and throw AssertionError if not okey. Clean stack traces. Simplicity. Built on [is-kindof][]. | [homepage](https://github.com/tunnckocore/assert-kindof#readme "Check native type of value and throw AssertionError if not okey. Clean stack traces. Simplicity. Built on [is-kindof][].")
- [clean-stacktrace](https://www.npmjs.com/package/clean-stacktrace): Clean up error stack traces - just a fork of [clean-stack][]. Working on node.js v0.10 and above. | [homepage](https://github.com/tunnckocore/clean-stacktrace#readme "Clean up error stack traces - just a fork of [clean-stack][]. Working on node.js v0.10 and above.")
- [is-async-function](https://www.npmjs.com/package/is-async-function): Is function really asynchronous function? Trying to guess that based on check if [common-callback-names][] exists as function arguments names or… [more](https://github.com/tunnckocore/is-async-function#readme) | [homepage](https://github.com/tunnckocore/is-async-function#readme "Is function really asynchronous function? Trying to guess that based on check if [common-callback-names][] exists as function arguments names or you can pass your custom.")
- [kind-of-extra](https://www.npmjs.com/package/kind-of-extra): Additional functionality to [kind-of][] type check utility. Support promises, generators, streams, errors. | [homepage](https://github.com/tunnckocore/kind-of-extra#readme "Additional functionality to [kind-of][] type check utility. Support promises, generators, streams, errors.")
- [minibase](https://www.npmjs.com/package/minibase): Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the already existing… [more](https://github.com/node-minibase/minibase#readme) | [homepage](https://github.com/node-minibase/minibase#readme "Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the already existing [base][] plugins.")
- [redolent](https://www.npmjs.com/package/redolent): Simple promisify with sane defaults, works on node 0.10 if you provide custom Promise through options | [homepage](https://github.com/hybridables/redolent#readme "Simple promisify with sane defaults, works on node 0.10 if you provide custom Promise through options")
- [try-catch-core](https://www.npmjs.com/package/try-catch-core): Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful for and… [more](https://github.com/hybridables/try-catch-core#readme) | [homepage](https://github.com/hybridables/try-catch-core#readme "Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful for and used in higher-level libs such as [always-done][] to handle completion of anything.")

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/tunnckoCore/clean-stacktrace-relative-paths/issues/new).  
Please read the [contributing guidelines](CONTRIBUTING.md) for advice on opening issues, pull requests, and coding standards.  
If you need some help and can spent some cash, feel free to [contact me at CodeMentor.io](https://www.codementor.io/tunnckocore?utm_source=github&utm_medium=button&utm_term=tunnckocore&utm_campaign=github) too.

**In short:** If you want to contribute to that project, please follow these things

1. Please DO NOT edit [README.md](README.md), [CHANGELOG.md](CHANGELOG.md) and [.verb.md](.verb.md) files. See ["Building docs"](#building-docs) section.
2. Ensure anything is okey by installing the dependencies and run the tests. See ["Running tests"](#running-tests) section.
3. Always use `npm run commit` to commit changes instead of `git commit`, because it is interactive and user-friendly. It uses [commitizen][] behind the scenes, which follows Conventional Changelog idealogy.
4. Do NOT bump the version in package.json. For that we use `npm run release`, which is [standard-version][] and follows Conventional Changelog idealogy.

Thanks a lot! :)

## Building docs
Documentation and that readme is generated using [verb-generate-readme][], which is a [verb][] generator, so you need to install both of them and then run `verb` command like that

```
$ npm install verbose/verb#dev verb-generate-readme --global && verb
```

_Please don't edit the README directly. Any changes to the readme must be made in [.verb.md](.verb.md)._

## Running tests
Clone repository and run the following in that cloned directory

```
$ npm install && npm test
```

## Author
**Charlike Mike Reagent**

+ [github/tunnckoCore](https://github.com/tunnckoCore)
+ [twitter/tunnckoCore](https://twitter.com/tunnckoCore)
+ [codementor/tunnckoCore](https://codementor.io/tunnckoCore)

## License
Copyright © 2017, [Charlike Mike Reagent](https://i.am.charlike.online). Released under the [MIT License](LICENSE).

***

_This file was generated by [verb-generate-readme](https://github.com/verbose/verb-generate-readme), v0.4.3, on March 12, 2017._  
_Project scaffolded using [charlike][] cli._

[always-done]: https://github.com/hybridables/always-done
[async-done]: https://github.com/gulpjs/async-done
[base]: https://github.com/node-base/base
[charlike]: https://github.com/tunnckocore/charlike
[clean-stack]: https://github.com/sindresorhus/clean-stack
[clean-stacktrace]: https://github.com/tunnckocore/clean-stacktrace
[commitizen]: https://github.com/commitizen/cz-cli
[common-callback-names]: https://github.com/tunnckocore/common-callback-names
[dezalgo]: https://github.com/npm/dezalgo
[is-kindof]: https://github.com/tunnckocore/is-kindof
[kind-of]: https://github.com/jonschlinkert/kind-of
[once]: https://github.com/isaacs/once
[standard-version]: https://github.com/conventional-changelog/standard-version
[verb-generate-readme]: https://github.com/verbose/verb-generate-readme
[verb]: https://github.com/verbose/verb

[license-url]: https://www.npmjs.com/package/clean-stacktrace-relative-paths
[license-img]: https://img.shields.io/npm/l/clean-stacktrace-relative-paths.svg

[downloads-url]: https://www.npmjs.com/package/clean-stacktrace-relative-paths
[downloads-img]: https://img.shields.io/npm/dt/clean-stacktrace-relative-paths.svg

[codeclimate-url]: https://codeclimate.com/github/tunnckoCore/clean-stacktrace-relative-paths
[codeclimate-img]: https://img.shields.io/codeclimate/github/tunnckoCore/clean-stacktrace-relative-paths.svg

[travis-url]: https://travis-ci.org/tunnckoCore/clean-stacktrace-relative-paths
[travis-img]: https://img.shields.io/travis/tunnckoCore/clean-stacktrace-relative-paths/master.svg?label=linux

[appveyor-url]: https://ci.appveyor.com/project/tunnckoCore/clean-stacktrace-relative-paths
[appveyor-img]: https://img.shields.io/appveyor/ci/tunnckoCore/clean-stacktrace-relative-paths/master.svg?label=windows

[coverage-url]: https://codecov.io/gh/tunnckoCore/clean-stacktrace-relative-paths
[coverage-img]: https://img.shields.io/codecov/c/github/tunnckoCore/clean-stacktrace-relative-paths/master.svg

[david-url]: https://david-dm.org/tunnckoCore/clean-stacktrace-relative-paths
[david-img]: https://img.shields.io/david/tunnckoCore/clean-stacktrace-relative-paths.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg

[paypalme-url]: https://www.paypal.me/tunnckoCore
[paypalme-img]: https://img.shields.io/badge/paypal-donate-brightgreen.svg



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.4"></a>
## [1.0.4](https://github.com/tunnckocore/clean-stacktrace-relative-paths/compare/v1.0.3...v1.0.4) (2017-03-17)


### Bug Fixes

* **tests:** typos, because fast fingers ([828f56d](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/828f56d)), closes [#3](https://github.com/tunnckocore/clean-stacktrace-relative-paths/issues/3)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/tunnckocore/clean-stacktrace-relative-paths/compare/v1.0.2...v1.0.3) (2017-03-17)


### Bug Fixes

* **package:** bump mukla testing ([a7e8eb0](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/a7e8eb0))
* **regex:** improve ([f7e5945](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/f7e5945))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/tunnckocore/clean-stacktrace-relative-paths/compare/v1.0.1...v1.0.2) (2017-03-15)


### Bug Fixes

* **ci:** update tests for CI services ([3d75459](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/3d75459)), closes [#2](https://github.com/tunnckocore/clean-stacktrace-relative-paths/issues/2)
* **tests:** typo ([82ae24c](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/82ae24c))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/tunnckocore/clean-stacktrace-relative-paths/compare/v1.0.0...v1.0.1) (2017-03-15)


### Bug Fixes

* **regex:** improve regex ([5bae1d7](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/5bae1d7))



<a name="1.0.0"></a>
# 1.0.0 (2017-03-12)


### Bug Fixes

* **ci:** test on node < 4 ([2a0f909](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/2a0f909))
* **package:** force LOACK VERSIONB! GK! ([02b26b3](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/02b26b3))
* **readme:** tweaks ([3b50341](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/3b50341))
* **tests:** damn windows ([5b0150b](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/5b0150b))


### Chores

* **initial:** scaffold ([7b1233e](https://github.com/tunnckocore/clean-stacktrace-relative-paths/commit/7b1233e))


### BREAKING CHANGES

* **initial:** First release





## 0.0.0 - 2017-03-11
- Initial commit

test.js
/*!
 * clean-stacktrace-relative-paths <https://github.com/tunnckoCore/clean-stacktrace-relative-paths>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

var path = require('path')
var isCI = require('is-ci')
var test = require('mukla')
var cleanStack = require('clean-stacktrace')
var relativePaths = require('./index')

test('should export a function that returns a mapper function', function (done) {
  var mapper = relativePaths()
  test.strictEqual(typeof mapper, 'function')
  done()
})

test('should allow passing custom cwd', function (done) {
  var er = new Error('abc error foo')
  var mapper = relativePaths('../qux')
  var stack = cleanStack(er.stack, mapper)

  var fp = '..' + path.sep + 'clean-stacktrace-relative-paths' + path.sep + 'test.js'
  test.strictEqual(stack.indexOf(fp) > 5, true)
  done()
})

test('should be used as `mapper` function to `clean-stacktrace`', function (done) {
  var error = new Error('fixture err')
  var stack = cleanStack(error.stack, relativePaths())
  test.ok(/\(test\.js:/.test(stack))
  done()
})

test('should work for paths non in parens', function qxu (done) {
  var line = 'at Function.qxu (/home/charlike/apps/clean-stacktrace-relative-paths/test.js:44:15)'
  var relative = relativePaths()
  var res = relative(line)

  if (isCI) {
    test.strictEqual(/at Function\.qxu/.test(res), true)
    test.strictEqual(/\.\./.test(res), true)
    test.strictEqual(/test\.js:44:15/.test(res), true)
  } else {
    test.strictEqual(res, 'at Function.qxu (test.js:44:15)')
  }
  done()
})

test('should work for lines like "at /full/absolute/path.js:3:1" with no "place"', function (done) {
  var line = 'at /home/charlike/apps/clean-stacktrace-relative-paths/test.js:3:1'
  var res = relativePaths()(line)

  if (isCI) {
    test.strictEqual(/at \.\./i.test(res), true)
    test.strictEqual(/test\.js:3:1/i.test(res), true)
  } else {
    test.strictEqual(res, 'at test.js:3:1')
  }
  done()
})


index.js
/*!
 * clean-stacktrace-relative-paths <https://github.com/tunnckoCore/clean-stacktrace-relative-paths>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (https://i.am.charlike.online)
 * Released under the MIT license.
 */

'use strict'

var path = require('path')

/**
 * > Cleans stack traces to use relative paths,
 * instead of absolute paths. Pass a `cwd` if you want
 * to customize, otherwise `process.cwd()` is used by default.
 *
 * Meant to be used as `mapper` argument in [clean-stacktrace][]
 *
 * **Example**
 *
 * ```js
 * const cleanStack = require('clean-stacktrace')
 * const relativePaths = require('clean-stacktrace-relative-paths')
 *
 * const error = new Error('Missing unicorn')
 * console.log(error.stack)
 * // =>
 * // Error: Missing unicorn
 * //     at Object.<anonymous> (/Users/sindresorhus/dev/clean-stack/unicorn.js:2:15)
 * //     at Module._compile (module.js:409:26)
 * //     at Object.Module._extensions..js (module.js:416:10)
 * //     at Module.load (module.js:343:32)
 * //     at Function.Module._load (module.js:300:12)
 * //     at Function.Module.runMain (module.js:441:10)
 * //     at startup (node.js:139:18)
 *
 * const mapper = relativePaths()
 * const stack = cleanStack(error.stack, mapper)
 * console.log(stack)
 * // =>
 * // Error: Missing unicorn
 * //     at Object.<anonymous> (unicorn.js:2:15)
 * ```
 *
 * @name   relativePaths
 * @param  {String} `[cwd]` custom current working directory
 * @return {Function} a function that can be passed as `mapper` to [clean-stacktrace][]
 * @api public
 */

module.exports = function relativePaths (cwd) {
  return function mapper (line) {
    var m = /at\s+([^\s/]+)?\s?(.*)$/.exec(line)
    if (m) {
      /* istanbul ignore next */
      var filepath = (m[2] && m[2].length ? m[2] : m[1])
      filepath = filepath.replace(/^\(/, '').replace(/\)$/, '')

      var relative = path.relative(cwd || process.cwd(), filepath)
      return line.replace(filepath, relative)
    }

    return line
  }
}
